lib/p2p/packets/Packet.ts

```diff
@@ -1,11 +1,36 @@
 import PacketType from './PacketType';
+import CryptoJS from 'crypto-js';
+import SHA256 from 'crypto-js/sha256';
+
+type PacketHeader = {
+  hash?: string;
+  reqHash?: string;
+};
+
+enum MessageType {
+  UNILATERAL,
+  REQUEST,
+  RESPONSE,
+}
 
 abstract class Packet {
+  public header: PacketHeader;
   public abstract type: PacketType;
   public abstract body: any;
-  public abstract responseType?: PacketType;
-  public abstract responseTimeout?: number;
-  public abstract toRaw(): string;
+  public abstract messageType: MessageType;
+
+  constructor (body: any, header?: PacketHeader) {
+    this.header = {
+      hash: (header && header.hash) || SHA256(JSON.stringify(body)).toString(CryptoJS.enc.Hex),
```

Comment:
Another minor thing, maybe we should use Base64 encoding instead of hex? Fewer characters, and therefore bytes, that need to be transferred and stored in memory that way. Human readability is not a real concern for packet headers as far as I can tell.

Reply:
Fixed. We can also use a different hash function.

Reply:
True... from some quick research it looks like MD5 is considerably faster, and since we'll be calculating these hashes very frequently it's probably worth switching. It has some known collision vulnerabilities but since a collision here would not be catastrophic, I think it's good enough for our purposes.

Reply:
Yes, that what I was thinking, just started with SHA256 as a default.